runtime-modules/storage/src/lib.rs

```diff
@@ -3942,4 +3602,342 @@ impl<T: Trait> Module<T> {
+                    .ok()
+            })
+            // attempt to update bucket voucher
+            .map(|(id, bk)| {
+                bk.voucher
+                    .clone()
+                    .update(new_voucher_update)
+                    .map_or_else(
+                        |e| match e {
+                            "size" => Err(Error::<T>::StorageBucketObjectSizeLimitReached),
+                            _ => Err(Error::<T>::StorageBucketObjectNumberLimitReached),
+                        },
+                        |voucher| Ok((id, voucher)),
+                    )
+                    .map(|(id, voucher)| (id, StorageBucket::<T> { voucher, ..bk }))
+            });
+
+        match op {
+            BagOperationParams::<T>::Create(..) => {
+                let correct = tmp
+                    .filter_map(|r| {
+                        r.map(|(id, mut bk)| {
+                            bk.register_bag_assignment();
+                            (id, bk)
+                        })
+                        .ok()
+                    })
+                    .collect::<BTreeMap<_, _>>();
+                ensure!(
+                    !correct.is_empty(),
+                    Error::<T>::StorageBucketIdCollectionsAreEmpty
+                );
+                Ok(correct)
+            }
+            BagOperationParams::<T>::Update(..) => tmp.collect(),
+            BagOperationParams::<T>::Delete => tmp
+                .map(|r| {
+                    r.map(|(id, mut bk)| {
+                        bk.unregister_bag_assignment();
+                        (id, bk)
+                    })
+                })
+                .collect(),
+        }
+    }
+
+    fn generate_new_distribution_buckets(
+        bucket_ids: &BTreeSet<DistributionBucketId<T>>,
+        op: &BagOperationParams<T>,
+    ) -> BTreeMap<DistributionBucketId<T>, DistributionBucket<T>> {
+        bucket_ids
+            .iter()
+            .filter_map(|id| {
+                Self::ensure_distribution_bucket_exists(id)
+                    .map(|mut bk| match op {
+                        BagOperationParams::<T>::Create(..) => {
+                            bk.register_bag_assignment();
+                            (id.clone(), bk)
+                        }
+                        BagOperationParams::<T>::Delete => {
+                            bk.unregister_bag_assignment();
+                            (id.clone(), bk)
+                        }
+                        _ => (id.clone(), bk), // no op in case of objects upload/delete operation
+                    })
+                    .ok()
+            })
+            .collect()
+    }
+
+    fn compute_net_prize(
+        init_value: NetDeletionPrize<T>,
+        num_obj_to_create: usize,
+        num_obj_to_delete: usize,
+    ) -> NetDeletionPrize<T> {
+        let amnt = T::DataObjectDeletionPrize::get();
+        init_value
+            .add_balance(
+                iter::repeat(amnt)
```

Comment:
Can't we multiply here?

Reply:
`BalanceOf<T> * usize` multiplication does not work, so I didn't change the accounting method for the deletion prize (meaning even before my refactoring `iter::repeat + fold` was used instead of simple multiplication.

Reply:
`BalanceOf<T>` is defined on runtime + `usize` is architecture-dependent [see here](https://doc.rust-lang.org/nightly/std/primitive.usize.html). So even converting both operands into say `u64` multiply and then convert the result back to `BalanceOf<T>` is not feasible IMHO

Reply:
:+1:

Reply:
You can cast usize to Balance instead.

This code compiles:
```
        let temp: BalanceOf::<T> = num_obj_to_create.saturated_into();
        let result = temp * amnt;